### PR TITLE
[Backend] Partial API Implementation 📄 

### LINKED ISSUE #45: [Backend] SQL & API: Add sql for admin / buyer / general
Add sql for the API under `admin.go` and `buyer.go` and `general.go`

### PR DESCRIPTION
**Pull request by Error0229**
Resolve #45 
### Important Notes 
Everything things under are strongly debatable, feel free to recommend a possible/better approach.
+ The current `News` implementation.
+ The current `cart` with `coupon` logics
+ Using `"image_id" AS "image_url"` while retrieving rows in `sql` for better semantics in the API returning struct.
+ The current file structure
+ Don't merge a PR big like this
+ Implement some test for this PR

### Changes
+ Refactor file structure
+ Add database analysis for `sqlc`
+ Add migration for searching 
+ Set `QUERY_LIMIT = 20`  since 10 is too small
+ Add empty `fileName` string check for `mc.GetFileURL`
---
Implement tons of APIs
The checked APIs are well-implemented, and subjected to minimal testing.
- [x] `GET "/admin/user" adminGetUser`
- [x] `DELETE "/admin/user/:username" adminDisableUser`
- [x] `GET "/admin/coupon" adminGetCoupon`
- [x] `GET "/admin/coupon/:id" adminGetCouponDetail`
- [x] `POST "/admin/coupon" adminAddCoupon`
- [x] `PATCH "/admin/coupon/:id" adminEditCoupon`
- [x] `DELETE "/admin/coupon/:id" adminDeleteCoupon`
- [x] `GET "/admin/report" adminGetReport`
---
- [x] `GET "/shop/:seller_name" getShopInfo`
- [x] `GET "/shop/:seller_name/coupon" getShopCoupon`
- [x] `GET "/tag/:id" getTagInfo`
- [x] `GET "/product/:id" getProductInfo`
- [x] `GET "/discover" getDiscover`
- [x] `GET "/news" getNews`
- [x] `GET "/news/:id" getNewsDetail`
- [x] `GET "/shop/:seller_name/search" searchShopProduct`
- [x] `GET "/search" search // search both product and shop`
- [x] `GET "/search/shop" searchShopByName`
---
- [x] `GET "/buyer/order" buyerGetOrderHistory`
- [x] `GET "/buyer/order/:id" buyerGetOrderDetail`
- [x] `PATCH "/buyer/order/:id" UpdateOrderStatus` 
- [x] `GET "/buyer/cart" buyerGetCart`
- [x] `POST "/buyer/cart/product/:id" buyerAddProductToCart // since the cart might not exist yet`
- [x] `GET "/buyer/cart/:id/coupon" buyerGetCoupon`
- [x] `POST "/buyer/cart/:cart_id/coupon/:coupon_id" buyerAddCouponToCart`
- [x] `PATCH "/buyer/cart/:cart_id/product/:product_id" buyerEditProductInCart`
- [x] `DELETE "/buyer/cart/:cart_id/product/:product_id" buyerDeleteProductFromCart `
- [x] `DELETE "/buyer/cart/:cart_id/coupon/:coupon_id" buyerDeleteCouponFromCart`
- [x] `GET "/buyer/cart/:id/checkout" buyerGetCheckout`
- [x] `POST "/buyer/cart/:id/checkout" buyerCheckout`

### Test approach
Since it's too complicated to write a full test flow for these APIs, I strongly recommend using Swagger to test them.
Also you might need to use another version of [`data.json`](https://gist.github.com/Error0229/c9d75c454822ff61a2ecc2be7aa43ab1) to test due to the fixed buyer name and also for richer and stable data.

### CODE REVIEW

File: pkg/router/general/shop.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+			logger.Errorw("failed to bind query parameter", "error", err)
+			return echo.NewHTTPError(http.StatusBadRequest)
+		}
+		if err := q.Validate(); err != nil {
+			logger.Errorw("failed to validate query parameter", "error", err)
+			return echo.NewHTTPError(http.StatusBadRequest, "query parameter is invalid")
+		}
+		shop_id, err := pg.Queries.ShopExists(c.Request().Context(), seller_name)
+		if err != nil {
+			if err == pgx.ErrNoRows {
```
Comment (NOOBDY) [inline pkg/router/general/shop.go]:
```suggestion
			if errors.Is(err, pgx.ErrNoRows) {
```

**Review by NOOBDY — CHANGES_REQUESTED** [review]

File: db/queries/admin.sql [hunk 2]
```diff
@@ -142,21 +162,54 @@ WHERE
...
+    S."image_id" AS "image_url",
+    SUM(O."total_price") AS "total_sales"
+FROM
+    "shop" AS S,
+    "order_history" AS O
+WHERE
+    S."id" = O."shop_id"
+    AND O."status" = 'paid'
+    AND O."created_at" < CAST((CAST(@date::TEXT AS TIMESTAMPTZ) +(INTERVAL '1 month')) AS TIMESTAMPTZ)
+    AND O."created_at" >= CAST(@date::TEXT AS TIMESTAMPTZ)
```
Comment (x06lan) [inline db/queries/admin.sql]: input timestamptz will be better

File: pkg/router/admin/coupon.go [hunk 3]
```diff
@@ -142,13 +83,22 @@ func adminGetCouponDetail(pg *db.DB, logger *zap.SugaredLogger) echo.HandlerFunc
...
 		}
+		discount, err := coupon.Discount.Float64Value()
+		if err != nil {
+			logger.Errorw("failed to parse discount", "error", err)
+			return echo.NewHTTPError(http.StatusBadRequest, "Invalid discount")
+		}
+		if discount.Float64 < 0 || discount.Float64 > 100 {
+			logger.Errorw("discount is invalid", "discount", discount)
+			return echo.NewHTTPError(http.StatusBadRequest, "Discount is invalid")
+		}
```
Comment (x06lan) [inline pkg/router/admin/coupon.go]: only percentage need to check discount > 100

File: pkg/router/admin/coupon.go [hunk 4]
```diff
@@ -179,13 +129,22 @@ type PrettierCoupon struct { // for swagger
...
```
Comment (x06lan) [inline pkg/router/admin/coupon.go]: same

File: pkg/router/admin/coupon.go [hunk 5]
```diff
@@ -123,12 +64,12 @@ func adminGetCouponDetail(pg *db.DB, logger *zap.SugaredLogger) echo.HandlerFunc
 			logger.Infow("coupon not found", "id", id)
```
Comment (x06lan) [inline pkg/router/admin/coupon.go]: change to errorw

File: pkg/router/buyer/checkout.go [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+	Shipment      int32            `json:"shipment"`
+	TotalDiscount int32            `json:"total_discount"`
+	Coupons       []couponDiscount `json:"coupons"`
+	Total         int32            `json:"total"`
+	Payments      json.RawMessage  `json:"payments"`
+}
+
+func getShipmentFee(total int32) int32 {
+	return int32(10 * math.Log(100*float64(total)))
+}
```
Comment (x06lan) [inline pkg/router/buyer/checkout.go:35]: why have log in there?

File: db/queries/buyer.sql [hunk 7]
```diff
@@ -28,10 +83,515 @@ WHERE
...
+WITH delete_expire_coupons AS (
+    DELETE FROM "cart_coupon" AS CC USING "coupon" AS CO, "cart" AS C, "user" AS U
+WHERE U."username" = $1
+    AND C."user_id" = U."id"
+    AND C."id" = CC."cart_id"
+    AND C."id" = @cart_id
+    AND CC."coupon_id" = CO."id"
+    AND NOW() > CO."expire_date"
+)
+SELECT
```
Comment (x06lan) [inline db/queries/buyer.sql:438]: dose this mean if there only have one product are valid then all whole cart_product is valid?

File: pkg/router/buyer/coupon.go [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+	return t
+}
+func (t *tagSet) Intersect(other *tagSet) bool {
+	for k := range t.Set {
+		if other.Set[k] {
+			return true
+		}
+	}
+	return false
+}
```
Comment (x06lan) [inline pkg/router/buyer/coupon.go]: maybe your can make this as single file

File: db/queries/buyer.sql [hunk 9]
```diff
@@ -28,10 +83,515 @@ WHERE
...
+-- name: ValidatePayment :one
+SELECT
+    EXISTS (
+        SELECT
+            1
+        FROM
+            "user"
+        WHERE
+            "username" = $1
+            AND ("credit_card" = $2
```
Comment (x06lan) [inline db/queries/buyer.sql:601]: why have random in there?

File: db/init_db.go [hunk 10]
```diff
@@ -27,6 +28,13 @@ func NewDB() (*DB, error) {
 
 	return &DB{Queries: q, pool: pool}, nil
 }
+func (db *DB) NewTx(ctx context.Context) (pgx.Tx, error) {
+	tx, err := db.pool.Begin(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return tx, nil
```
Comment (NOOBDY) [inline db/init_db.go]:
```suggestion
	return db.pool.Begin(ctx)
```
Upon inspecting the source code I noticed `pool.Begin()` already returns `nil` when the operation fails

File: pkg/router/buyer/checkout.go [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+		productTag := make(map[int32]*tagSet)
+		products, err := pg.Queries.GetProductFromCart(c.Request().Context(), cartID)
+		if err != nil {
+			logger.Errorw("failed to get product from cart", "error", err)
+			return echo.NewHTTPError(http.StatusInternalServerError)
+		}
+		// sort to make customer get most discount
+		sort.Slice(products, func(i, j int) bool {
+			return products[i].Price.Int.Cmp(products[j].Price.Int) > 0
+		})
```
Comment (NOOBDY) [inline pkg/router/buyer/checkout.go]: can `ORDER BY DESC` work in this case?

File: pkg/router/general/search.go [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+	common.QueryParams
+}
+
+func (q *searchParams) Validate() error {
+	if q.Q == "" {
+		return errors.New("search query is empty")
+	}
+	if q.MinPrice < 0 || q.MaxPrice < 0 || q.MinPrice > q.MaxPrice {
+		return errors.New("invalid price range")
+	}
```
Comment (x06lan) [inline pkg/router/general/search.go:32]: do user must set min_price and max_price?

File: test/db/main.go [hunk 13]
```diff
@@ -21,7 +21,7 @@ func main() {
 	defer pg.Close()
 
 	TestInsertData(pg)
-	TestDeleteData(pg)
+	// TestDeleteData(pg)
```
Comment (x06lan) [inline test/db/main.go]: remove comment

File: pkg/router/general/search.go [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+			Limit:      int32(q.Limit),
+			Query:      q.Q,
+			MinPrice: pgtype.Numeric{
+				Int:   big.NewInt(int64(q.MinPrice)),
+				Valid: true,
+			},
+			MaxPrice: pgtype.Numeric{
+				Int:   big.NewInt(int64(q.MaxPrice)),
+				Valid: true,
+			},
```
Comment (x06lan) [inline pkg/router/general/search.go]: miss the exp part

File: pkg/router/general/search.go [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+			"haveCoupon": &dbq.HasCoupon,
+		}
+		for k, v := range paramMap {
+			if c.QueryParam(k) == "" {
+				switch val := v.(type) {
+				case *pgtype.Numeric:
+					val.Valid = false
+				case *pgtype.Bool:
+					val.Valid = false
+				}
```
Comment (x06lan) [inline pkg/router/general/search.go]: why set valid to false?

File: pkg/router/general/search.go [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+			}
+		}
```
Comment (x06lan) [inline pkg/router/general/search.go]: ditto

File: pkg/router/admin/user.go [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+		username := c.Param("username")
+		if username == "" { // would not happen in future
+			logger.Errorw("username is empty")
+			return echo.NewHTTPError(http.StatusBadRequest)
+		}
+		if execRows, err := pg.Queries.DisableUser(c.Request().Context(), username); err != nil {
+			logger.Errorw("failed to disable user", "error", err)
+			return echo.NewHTTPError(http.StatusInternalServerError, "Failed to disable user")
+		} else if execRows == 0 {
+			logger.Infow("user not found", "username", username)
```
Comment (x06lan) [inline pkg/router/admin/user.go]: errorw

**Review by NOOBDY — DISMISSED** [review]

**Review by x06lan — CHANGES_REQUESTED** [review]

File: pkg/router/buyer/checkout.go (hunk 6 above)
Comment (Error0229) [reply pkg/router/buyer/checkout.go:35]: why not if we can

File: pkg/router/general/search.go (hunk 12 above)
Comment (Error0229) [reply pkg/router/general/search.go:32]: since it default zero, this only check negative and min>max

**Review by NOOBDY — APPROVED** [review]

**Review by NOOBDY — APPROVED** [review]

**Review by x06lan — APPROVED** [review]

**Pull request merged**